### PR TITLE
Refactor build.ts into multiple files

### PR DESCRIPTION
**Pull request by mads-hartmann**
THIS IS STILL IN DRAFT

## Description

This PR splits `build.ts` into smaller files so that `build.ts` becomes a fairly small file where the overall sequence and structure of the build and be easily understood.

The structure now follows this pattern

```
.werft
    jobs
        build
            build-and-publish.ts
            deploy-to-preview-environment.ts
            ... the other "phases" of the build has their own files
    build.yaml
    build.ts
```

and `build.ts` primarily looks like this

```typescript
async function run(context: any) {
    const config = jobConfig(werft, context)

    await validateChanges(werft)
    await prepare(werft)
    await buildAndPublish(werft, config)
    await coverage(werft, config)

    if (config.noPreview) {
        werft.phase("deploy", "not deploying");
        console.log("no-preview or publish-release is set");
        return
    }

    await deployToPreviewEnvironment(werft, config)
    await triggerIntegrationTests(werft, config, context.Owner)
}

```

I have tried to clean up as little as possible in this PR but I have made the following changes

1. I have consolidated the "configuration" of the job into a JobConfig type. The idea is that we compute this in the beginning of the job based on Werft attributes and other details, and then just pass that around. While that does mean we're passing a rather big object around to most functions, it means at least we do our parsing and interpretation of the "configuration" at a single place.
2. I have not tried to merge it with DeployConfig so right now the two objects share a lot of the same fields. I wanted to keep this PR as small as possible.

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/8065

## How to test

- [ ] Preview environment using Helm (and observability) should work
- [ ] Preview environment using Installer should work
- [ ] Preview environment using VM should work

## Release Notes

```release-note
NONE
```

## Documentation

N/A

### CODE REVIEW

File: .werft/jobs/build/build-and-publish.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    if (withContrib || publishRelease) {
+        exec(`leeway build --docker-build-options network=host --werft=true -c remote ${dontTest ? '--dont-test' : ''} -Dversion=${version} -DimageRepoBase=${imageRepo} contrib:all`);
+    }
+    exec(`leeway build --docker-build-options network=host --werft=true -c remote ${dontTest ? '--dont-test' : ''} ${retag} --coverage-output-path=${coverageOutput} -Dversion=${version} -DremoveSources=false -DimageRepoBase=${imageRepo} -DlocalAppVersion=${localAppVersion} -DSEGMENT_IO_TOKEN=${process.env.SEGMENT_IO_TOKEN} -DnpmPublishTrigger=${publishToNpm ? Date.now() : 'false'} -DjbMarketplacePublishTrigger=${publishToJBMarketplace ? Date.now() : 'false'}`);
+    if (publishRelease) {
+        try {
+            werft.phase("publish", "checking version semver compliance...");
+            if (!semver.valid(version)) {
+                // make this an explicit error as early as possible. Is required by helm Charts.yaml/version
+                throw new Error(`'${version}' is not semver compliant and thus cannot used for Self-Hosted releases!`)
```
Comment (ArthurSens) [inline .werft/jobs/build/build-and-publish.ts:43]:
```suggestion
                throw new Error(`'${version}' is not semver compliant and thus cannot be used for Self-Hosted releases!`)
```

File: .werft/jobs/build/job-config.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
+import { config } from "process";
+import { exec } from "../../util/shell";
+import { Werft } from "../../util/werft";
+
+export interface JobConfig {
```
Comment (ArthurSens) [inline .werft/jobs/build/job-config.ts:5]: Super nice to see the usage of interfaces! Our CI can finally look like a proper software :)

File: .werft/jobs/build/job-config.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    analytics: string
+    buildConfig: any;
+    cleanSlateDeployment: boolean
+    coverageOutput: string
+    dontTest: boolean;
+    dynamicCPULimits: boolean;
+    installEELicense: boolean
+    localAppVersion: string
+    mainBuild: any;
```
Comment (ArthurSens) [inline .werft/jobs/build/job-config.ts:14]:
```suggestion
    mainBuild: boolean;
```

Could that be done? Is there any place where we don't use this as a boolean?

**Review by ArthurSens — COMMENTED** [review]
I know it's still a draft, I'm just taking a look over the changes since I was mentally preparing to tackle this one today 😂 

One question that was stuck on my mind this week, how do propagate tracing context when call a werft job from inside another?

**Comment by mads-hartmann** [comment]
> One question that was stuck on my mind this week, how do propagate tracing context when call a werft job from inside another?

We decided against using separate jobs as there's a performance overhead to creating new jobs. As the purpose of this work is to make the job a bit more maintainable and easier to understand it wasn't a strict requirement to separate it into different jobs, so we went with just separating it into different ts files instead of keeping build.ts nice and clean ☺️

However if we do ever separate it into multiple jobs we would have to propagate the trace headers as attributes to the Werft job and extend our instrumentation to pick those up when initializing the tracer ☺️

File: .werft/jobs/build/job-config.ts (hunk 3 above)
Comment (mads-hartmann) [reply .werft/jobs/build/job-config.ts:14]: I can give it a go - right now the types are based on what TypeScript interpreted the types to be ☺️

File: .werft/jobs/build/build-and-publish.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    exec(`leeway build --docker-build-options network=host --werft=true -c remote ${dontTest ? '--dont-test' : ''} --dont-retag --coverage-output-path=${coverageOutput} --save /tmp/dev.tar.gz -Dversion=${version} -DimageRepoBase=eu.gcr.io/gitpod-core-dev/dev dev:all`);
+
+    if (publishRelease) {
+        exec(`gcloud auth activate-service-account --key-file "/mnt/secrets/gcp-sa-release/service-account.json"`);
+    }
+    if (withContrib || publishRelease) {
+        exec(`leeway build --docker-build-options network=host --werft=true -c remote ${dontTest ? '--dont-test' : ''} -Dversion=${version} -DimageRepoBase=${imageRepo} contrib:all`);
+    }
+    exec(`leeway build --docker-build-options network=host --werft=true -c remote ${dontTest ? '--dont-test' : ''} ${retag} --coverage-output-path=${coverageOutput} -Dversion=${version} -DremoveSources=false -DimageRepoBase=${imageRepo} -DlocalAppVersion=${localAppVersion} -DSEGMENT_IO_TOKEN=${process.env.SEGMENT_IO_TOKEN} -DnpmPublishTrigger=${publishToNpm ? Date.now() : 'false'} -DjbMarketplacePublishTrigger=${publishToJBMarketplace ? Date.now() : 'false'}`);
+    if (publishRelease) {
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply .werft/jobs/build/build-and-publish.ts:38]: fair point!

**Comment by mads-hartmann** [comment]
Closing as I have to make the branch name shorts 😝 Opening up a new branch and PR

**Comment by mads-hartmann** [comment]
New PR here https://github.com/gitpod-io/gitpod/pull/8240

**Pull request closed without merging**